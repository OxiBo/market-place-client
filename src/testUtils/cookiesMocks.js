import { Cookies } from "react-cookie";
export const setMyCookies = (token, type, name, id) => {
  const cookiesMocks = new Cookies();
  cookiesMocks.set("token", token);
  cookiesMocks.set("type", type);
  cookiesMocks.set("name", name);
  cookiesMocks.set("id", id);
  return cookiesMocks;
};
