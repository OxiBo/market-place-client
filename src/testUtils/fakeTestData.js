import casual from "casual";

const fakeBuyer = () => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  age: 21,
  image:
    "https://res.cloudinary.com/di0hg10hd/image/upload/v1597378083/sickfits/qekc7yfql14ligautxzq.jpg",
  orderItems: [],
  reviews: [],
  type: "BUYER",
});

export { fakeBuyer };
