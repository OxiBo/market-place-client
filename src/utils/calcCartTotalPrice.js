const calcCartTotalPrice = (items) => {
  return items.reduce((sum, item) => {
    console.log(item)
    if (!item.product) return sum;
    return sum + item.count * item.product.price;
  }, 0);
};

export default calcCartTotalPrice;
