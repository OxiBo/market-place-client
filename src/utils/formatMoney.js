const formatMoney = (num) => {
    return num / 100;
  }

export default formatMoney