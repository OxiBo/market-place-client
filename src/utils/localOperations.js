import gql from "graphql-tag";

export const CART_OPEN_QUERY = gql`
  query {
    cartOpen @client
  }
`;


export const TOGGLE_CART_MUTATION = gql`
mutation TODDLE_CART_MUTATION{
    toggleCart @client{
        cartOpen
    }
}`