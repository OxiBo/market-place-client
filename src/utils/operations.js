import gql from "graphql-tag";
import { perPage } from "../configVars";

export const PRODUCT_SUBSCRIPTION = gql`
  subscription {
    product {
      node {
        id
        name
        description
        price
        rating
        stock
        seller {
          name
          id
        }
      }
      mutation
      previousValues {
        id
      }
    }
  }
`;

export const FETCH_PRODUCTS = gql`
query FETCH_PRODUCTS($skip: Int = 0, $first: Int = ${perPage} )
  {
    products(skip: $skip, first: $first, orderBy: createdAt_DESC ) {
      id
      name
      # description
      image
      price
      rating
      stock
      seller {
        id
        name
      }
    }
  }
`;

export const FETCH_SINGLE_PRODUCT = gql`
  query FETCH_SINGLE_PRODUCT($id: ID!) {
    product(where: { id: $id }) {
      id
      name
      department
      description
      largeImage
      price
      rating
      stock
      seller {
        id
        name
      }
      reviews {
        text
        rating
      }
    }
  }
`;

export const CREATE_PRODUCT = gql`
  mutation createProduct($data: CreateProductInput!) {
    createProduct(data: $data) {
      id
      name
      department
      description
      image
      largeImage
      price
      rating
      stock
    }
  }
`;

export const DELETE_PRODUCT = gql`
  mutation deleteProduct($id: ID!) {
    deleteProduct(id: $id) {
      id
    }
  }
`;

export const PAGINATION = gql`
  query PAGINATION {
    productsConnection {
      aggregate {
        count
      }
    }
  }
`;

export const LOGIN_USER = gql`
  mutation loginUser($data: LoginUserInput!) {
    loginUser(data: $data) {
      token
      user {
        id
        name
        age
        type
      }
    }
  }
`;

export const LOGIN_SELLER = gql`
  mutation loginSeller($data: LoginSellerInput!) {
    loginSeller(data: $data) {
      token
      seller {
        id
        name
        email
        type
      }
    }
  }
`;

export const SIGNUP_USER = gql`
  mutation createUser($data: CreateUserInput!) {
    createUser(data: $data) {
      token
      user {
        id
        name
        email
        age
        type
      }
    }
  }
`;

export const SIGNUP_SELLER = gql`
  mutation createSeller($data: CreateSellerInput!) {
    createSeller(data: $data) {
      token
      seller {
        id
        name
        email
        type
      }
    }
  }
`;

export const FETCH_USER_PROFILE = gql`
  query {
    meUser {
      id
      name
      email
      age
      type
    }
  }
`;

export const FETCH_SELLER_PROFILE = gql`
  query {
    meSeller {
      id
      name
      email
      type
    }
  }
`;

export const UPDATE_USER = gql`
  mutation updateUser($data: UpdateUserInput!) {
    updateUser(data: $data) {
      id
      name
      email
      age

      type
    }
  }
`;

export const UPDATE_SELLER = gql`
  mutation updateSeller($data: UpdateSellerInput!) {
    updateSeller(data: $data) {
      id
      name
      email

      type
    }
  }
`;
