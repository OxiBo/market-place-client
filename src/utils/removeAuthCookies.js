const removeAuthCookies = (cookies) => {
  cookies.remove("token", { path: "/" });
  cookies.remove("name", { path: "/" });
  cookies.remove("id", { path: "/" });
  cookies.remove("type", { path: "/" });
};

export default removeAuthCookies;
