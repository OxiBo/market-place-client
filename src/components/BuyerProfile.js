import React, { Component } from "react";
import { Query, Mutation } from "react-apollo";
import Helmet from "react-helmet";
import { Formik } from "formik";
import styled from "styled-components";
import { withCookies } from "react-cookie";
import Error from "./ErrorMessage";
import OrderedItemsList from "./OrderedItemsList";
import DeleteUser from "./DeleteUser";
import InnerContainer from "./styled/InnerContainer";
import Button from "./styled/Button";
import Buttons from "./styled/Buttons";
import Form from "./styled/Form";

import FormInput from "./styled/FormInput";
import uploadFile from "../utils/uploadFile";
import { FETCH_USER_PROFILE, UPDATE_USER } from "../utils/serverOperations";

import { noImage } from "../utils/utilVars";

const Content = styled(InnerContainer)`
  background-color: #fdfcfc;
  display: flex;
  div {
    margin: 0.5rem;
    padding: 1rem;
    h2 {
      font-size: 3rem;
    }
    p {
      padding: 0.5rem;
      span {
        text-transform: uppercase;
      }
    }
  }
  div.image {
    width: 25%;
    img {
      width: 100%;
    }
  }
`;

class BuyerProfile extends Component {
  state = {
    update: false,
    openOrderItems: false,
    role: this.props.cookies.get("type"),
  };

  render() {
    const { role } = this.state;

    return (
      <div>
        <Query query={FETCH_USER_PROFILE}>
          {({ data, loading, error }) => {
            if (loading) return <p>Loading....</p>;

            if (error) return <Error error={error} />;

            const {
              name: currentName,
              email: currentEmail,
              age: currentAge,
              image: currentImage,
            } = data.meUser;
            {
              /* console.log(data.meUser); */
            }
            return (
              <>
                <InnerContainer>
                  <Helmet>
                    <title>My Profile</title>
                  </Helmet>
                  <Content>
                    <div className="image">
                      <img src={currentImage || noImage} alt="currentName" />
                    </div>

                    <div data-test="buyer-profile">
                      {" "}
                      <h2>{currentName}</h2>
                      <p>
                        <span>Email:</span> {currentEmail}
                      </p>
                      {currentAge && (
                        <p>
                          <span>age:</span> {currentAge}
                        </p>
                      )}
                      {!this.state.update && (
                        <Button
                          data-test="update-profile-button"
                          onClick={() => this.setState({ update: true })}
                        >
                          Update Profile
                        </Button>
                      )}
                      <DeleteUser type={"BUYER"} cookies={this.props.cookies}/>
                   
                      <Button
                        data-test="ordered-items-button"
                        onClick={() =>
                          this.setState((prevState) => ({
                            openOrderItems: !prevState.openOrderItems,
                          }))
                        }
                      >
                        {this.state.openOrderItems ? "Hide" : "Show"} ordered
                        items
                      </Button>
                    </div>
                  </Content>
                  {this.state.update && (
                    <Mutation
                      mutation={UPDATE_USER}
                      refetchQueries={[{ query: FETCH_USER_PROFILE }]}
                    >
                      {(updateProfileMutation, { loading, error }) => {
                        return (
                          <>
                            {error && <Error error={error} />}
                            {loading && <div>Loading....</div>}

                            <Formik
                              initialValues={{
                                name: currentName,
                                email: currentEmail,
                                password: "",
                                confirmPassword: "",
                                image: "",
                                age: currentAge || null,
                              }}
                              validate={(values) => {
                                const {
                                  name,
                                  email,
                                  password,
                                  confirmPassword,
                                  age,
                                } = values;
                                const errors = {};

                                if (
                                  name &&
                                  (name.length < 2 || name.length > 15)
                                ) {
                                  errors.name =
                                    "Name cannot be less than 3 or more than 15 characters";
                                }

                                if (
                                  email &&
                                  !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(
                                    email
                                  )
                                ) {
                                  errors.email = "Invalid email address";
                                }

                                if (
                                  password &&
                                  (password.length < 8 || password.length > 20)
                                ) {
                                  errors.password =
                                    "Password has to be at least 8 characters long";
                                }

                                if (
                                  confirmPassword &&
                                  (confirmPassword.length < 8 ||
                                    confirmPassword.length > 20)
                                ) {
                                  errors.confirmPassword =
                                    "Password has to be at least 8 characters long";
                                }

                                if (password !== confirmPassword) {
                                  errors.confirmPassword =
                                    "Your passwords do not match";
                                }

                                if (
                                  !Number.isInteger(age) ||
                                  age < 14 ||
                                  age > 120
                                ) {
                                  errors.age =
                                    "Age has to be an integer not less then 14 or bigger then 120";
                                }
                                return errors;
                              }}
                              onSubmit={async (values, { setSubmitting }) => {
                                const updates = { ...values };
                                delete updates.confirmPassword;

                                // delete key value pairs which were not updated by the user
                                Object.entries(updates).forEach(
                                  ([key, val]) => {
                                    if (val === "") {
                                      delete updates[key];
                                    }
                                  }
                                );

                                try {
                                  const response = await updateProfileMutation({
                                    variables: { data: updates },
                                  }); // https://stackoverflow.com/questions/54574000/how-to-pass-variables-to-an-apollo-mutation-component
                                  console.log(response);
                                  this.setState({ update: false });
                                } catch (error) {
                                  console.error(error);
                                }
                                // if call setSubmit(false), the app is giving an error - https://github.com/formik/formik/issues/1449 - charleskoehl commented on Feb 11 â€¢

                                /* setSubmitting(false); */
                              }}
                            >
                              {({
                                values,
                                errors,
                                touched,
                                handleChange,
                                handleBlur,
                                handleSubmit,
                                isSubmitting,
                                setFieldValue,
                                /* and other goodies */
                              }) => {
                                return (
                                  <Form>
                                    <form onSubmit={handleSubmit}>
                                      <h2>Update Your Profile!</h2>
                                      <div className="fieldset">
                                        <div>
                                          <label htmlFor="name">Name</label>
                                          <div>
                                            <FormInput
                                              onChange={handleChange}
                                              onBlur={handleBlur}
                                              value={values.name}
                                              type="text"
                                              placeholder="Your name"
                                              name="name"
                                              id="name"
                                            />
                                            <p>
                                              {errors.name &&
                                                touched.name &&
                                                errors.name}
                                            </p>
                                          </div>
                                        </div>

                                        <div className="">
                                          <label htmlFor="file">Image:</label>
                                          <div>
                                            <FormInput
                                              type="file"
                                              id="file"
                                              name="file"
                                              placeholder="Upload an image"
                                              onChange={(e) =>
                                                uploadFile(e, setFieldValue)
                                              }
                                            />
                                            {values.image && (
                                              <img
                                                width="200"
                                                src={values.image}
                                                alt="Upload Preview"
                                              />
                                            )}
                                          </div>
                                        </div>

                                        <div>
                                          <label htmlFor="email">Email</label>
                                          <div>
                                            <FormInput
                                              onChange={handleChange}
                                              onBlur={handleBlur}
                                              value={values.email}
                                              type="text"
                                              name="email"
                                              id="email"
                                              placeholder="Enter your email"
                                            />
                                            <p>
                                              {errors.email &&
                                                touched.email &&
                                                errors.email}
                                            </p>
                                          </div>
                                        </div>
                                        <div>
                                          <label htmlFor="password">
                                            Password:
                                          </label>
                                          <div>
                                            <FormInput
                                              onChange={handleChange}
                                              onBlur={handleBlur}
                                              value={values.password}
                                              type="password"
                                              name="password"
                                              id="password"
                                              placeholder="Enter your password"
                                            />
                                            <p>
                                              {errors.password &&
                                                touched.password &&
                                                errors.password}
                                            </p>
                                          </div>
                                        </div>

                                        <div>
                                          <label htmlFor="confirmPassword">
                                            Confirm Password:
                                          </label>
                                          <div>
                                            <FormInput
                                              onChange={handleChange}
                                              onBlur={handleBlur}
                                              value={values.confirmPassword}
                                              type="password"
                                              name="confirmPassword"
                                              id="confirmPassword"
                                              placeholder="Confirm Your New Password"
                                            />
                                            <p>
                                              {errors.confirmPassword &&
                                                touched.confirmPassword &&
                                                errors.confirmPassword}
                                            </p>
                                          </div>
                                        </div>

                                        <div>
                                          <label htmlFor="age">Age:</label>
                                          <div>
                                            <FormInput
                                              onChange={handleChange}
                                              onBlur={handleBlur}
                                              value={values.age}
                                              type="number"
                                              name="age"
                                              id="age"
                                              placeholder="Enter your age"
                                            />
                                            <p>
                                              {errors.age &&
                                                touched.age &&
                                                errors.age}
                                            </p>
                                          </div>
                                        </div>
                                      </div>
                                      <Buttons>
                                        <Button
                                          onClick={() => {
                                            this.setState({ update: false });
                                          }}
                                        >
                                          Cancel
                                        </Button>
                                        <Button
                                          type="submit"
                                          disabled={isSubmitting}
                                        >
                                          Submit Changes
                                        </Button>
                                      </Buttons>
                                    </form>
                                  </Form>
                                );
                              }}
                            </Formik>
                          </>
                        );
                      }}
                    </Mutation>
                  )}
                  {this.state.openOrderItems && (
                    <OrderedItemsList userId={data.meUser.id} />
                  )}
                </InnerContainer>
              </>
            );
          }}
        </Query>
      </div>
    );
  }
}

export default withCookies(BuyerProfile);
export { BuyerProfile }; // ????
