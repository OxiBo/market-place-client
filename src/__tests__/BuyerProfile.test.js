import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { act } from "react-dom/test-utils";
import { ApolloConsumer } from "react-apollo";
import { MemoryRouter, withRouter } from "react-router-dom"; // https://html.developreference.com/article/13543175/React+Router+Invariant+Violation+test+fails+with+%3CLink%3E+outside+%3CRouter%3E+%5Bduplicate%5D
// why MemoryRouter - https://github.com/enzymejs/enzyme/issues/1112#issuecomment-447643271
import BuyerProfile from "../components/BuyerProfile";
// import { CURRENT_USER_QUERY } from "../components/User";
// import { MockedProvider } from "react-apollo/test-utils";
import { MockedProvider } from "@apollo/react-testing";
// mocking withCookies - https://github.com/reactivestack/cookies/blob/master/packages/react-cookie/src/__tests__/withCookies-test.js
import { CookiesProvider, withCookies, Cookies } from "react-cookie";
import { setMyCookies } from "../testUtils/cookiesMocks";
import {
  FETCH_USER_PROFILE,
  ALL_USER_ORDERITEMS_PAGINATION,
} from "../utils/serverOperations";
import { CART_OPEN_QUERY } from "../utils/localOperations";
import { fakeBuyer } from "../testUtils/fakeTestData";

const signedInBuyerProfile = [
  {
    request: { query: FETCH_USER_PROFILE },
    result: {
      data: {
        meUser: fakeBuyer(),
      },
    },
  },
];

const orderItemsPaginationMocks = (length) => {
  return [
    {
      request: {
        query: ALL_USER_ORDERITEMS_PAGINATION,
        variables: { userId: "4234" },
      },
      result: {
        data: {
          orderItemsConnection: {
            __typename: "aggregate",
            aggregate: {
              __typename: "count",
              count: length,
            },
          },
        },
      },
    },
  ];
};

let myCookies, wrapper, buyerProfile, pureComponent;
const Component = withCookies(BuyerProfile);
beforeEach(async() => {
  myCookies = setMyCookies(
    "token_lskdjflskdjflskdjflskdf",
    "BUYER",
    "Test",
    "lskdjflskdjflskdjflskdf"
  );

  wrapper = mount(
    <CookiesProvider cookies={myCookies}>
      <MemoryRouter>
        <MockedProvider
          mocks={[...signedInBuyerProfile, ...orderItemsPaginationMocks(7)]}
        >
          <Component />
        </MockedProvider>
      </MemoryRouter>
    </CookiesProvider>
  );
  await act(async () => wait());
  wrapper.update();
  buyerProfile = wrapper.find("BuyerProfile");
  pureComponent = buyerProfile.instance();
});

describe("<BuyerProfile/>", () => {
  it("renders with buyer logged in and matches snapshot", async () => {
    // console.log(buyerProfile.find("div[data-test='buyer-profile']").debug())
    expect(
      toJSON(buyerProfile.find("div[data-test='buyer-profile']"))
    ).toMatchSnapshot();
  });

  it('changes "show ordered items" button\'s text when clicked and matches snapshot', async () => {
    const orderItemsButton = wrapper.find(
      'button[data-test="ordered-items-button"]'
    );
    expect(orderItemsButton.text()).toContain("Show ordered items");
    pureComponent.setState({ openOrderItems: true });
    expect(orderItemsButton.text()).toContain("Hide ordered items");
    expect(toJSON(orderItemsButton)).toMatchSnapshot();
  });

  it('hides "update profile" button, when the button has been clicked and matches snapshot', async () => {
    const updateProfileButton = buyerProfile.find(
      'button[data-test="update-profile-button"]'
    );
    expect(updateProfileButton).toMatchSnapshot()
    pureComponent.setState({ update: true });
    expect(updateProfileButton).toEqual({});
  });


  
});
