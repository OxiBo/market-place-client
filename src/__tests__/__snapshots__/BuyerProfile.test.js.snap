// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BuyerProfile/> changes "show ordered items" button's text when clicked and matches snapshot 1`] = `
<button
  className="sc-AxhCb iCtNby"
  data-test="ordered-items-button"
  onClick={[Function]}
>
  Show
   ordered items
</button>
`;

exports[`<BuyerProfile/> hides "update profile" button, when the button has been clicked and matches snapshot 1`] = `ReactWrapper {}`;

exports[`<BuyerProfile/> renders with buyer logged in and matches snapshot 1`] = `
<div
  data-test="buyer-profile"
>
   
  <h2>
    Ms. Sydney Stehr
  </h2>
  <p>
    <span>
      Email:
    </span>
     
    Otto_Collier@Mueller.co.uk
  </p>
  <p>
    <span>
      age:
    </span>
     
    21
  </p>
  <styled.button
    data-test="update-profile-button"
    onClick={[Function]}
  >
    <button
      className="sc-AxhCb iCtNby"
      data-test="update-profile-button"
      onClick={[Function]}
    >
      Update Profile
    </button>
  </styled.button>
  <styled.button
    data-test="ordered-items-button"
    onClick={[Function]}
  >
    <button
      className="sc-AxhCb iCtNby"
      data-test="ordered-items-button"
      onClick={[Function]}
    >
      Show
       ordered items
    </button>
  </styled.button>
</div>
`;
